src/Std/System/Uri.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2019 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Lovett
+-/
+
+namespace System
+namespace Uri
+namespace UriEscape
+
+def hexDigitToNat? (c : Char) : Option Nat :=
+  if '0' ≤ c ∧ c ≤ '9' then some (c.toNat - '0'.toNat)
+  else if 'a' ≤ c ∧ c ≤ 'f' then some (c.toNat - 'a'.toNat + 10)
+  else if 'A' ≤ c ∧ c ≤ 'F' then some (c.toNat - 'A'.toNat + 10)
+  else none
+
+/-- Decode %HH escapings in the given string. Note that sometimes a consecutive
+sequence of multiple escapings can represet a utf-8 encoded sequence for
+a single unicode code point and these will also be decoded correctly. -/
+def decodeUri (uri : String) : String := Id.run do
+  let mut decoded : ByteArray := ByteArray.empty
+  let raw_bytes := uri.toUTF8
```

Comment:
```suggestion
  let rawBytes := uri.toUTF8
```
and elsewhere

Reply:
I'm afraid the mixing of raw byte access and `String.Iterator` has made the code quite messy. If you want to go that route, I'd suggest iterating over `rawBytes` directly, doing a byte-wise copy without involving any `String` functions at all. That should also be even faster.

Reply:
I guess it is safe to look for '%' since it can never be a valid 2nd, 3rd or 4th byte of a utf-8 encoding sequence according to https://en.wikipedia.org/wiki/UTF-8#Encoding.  But the byte right after '%' should be a hex digit and I think it is also safe that if it is a valid hex digit it cannot be the start of a new utf-8 encoding sequence, and same for the second hex digit... so I'll give it a try...

Reply:
what is the fastest way to get `bytes[i]` array access to the bytes in a ByteArray, the data is stored in `Array UInt8` and the data in an Array is stored as a List, so doesn't array access on ByteArray result in a nasty quadratic `List.get`... ?

Reply:
never mind perhaps `#eval ("😊".toUTF8)[2]!` is ok because it uses `uget` which is implemented externally in `lean_byte_array_uget`...

Reply:
Correct, the `List` does not actually exist unless you explicitly access it

Reply:
Ok, @Kha, it made no impact on performance, but I'm also running into what looks like a compiler bug, here's the repro:

```lean

def Zero : UInt8 := '0'.toNat.toUInt8
def Nine : UInt8 := '9'.toNat.toUInt8
def Lettera : UInt8 := 'a'.toNat.toUInt8
def Letterf : UInt8 := 'f'.toNat.toUInt8
def LetterA : UInt8 := 'A'.toNat.toUInt8
def LetterF : UInt8 := 'F'.toNat.toUInt8

/-- Decode %HH escapings in the given string. Note that sometimes a consecutive
sequence of multiple escapings can represet a utf-8 encoded sequence for
a single unicode code point and these will also be decoded correctly. -/
def decodeUri (uri : String) : String := Id.run do
  let mut decoded : ByteArray := ByteArray.empty
  let raw_bytes := uri.toUTF8
  let len := raw_bytes.size
  let mut i := 0
  let percent := '%'.toNat.toUInt8
  while i < len do
    let c := raw_bytes[i]!
    let (result, pos) := if c == percent && i + 1 < len then
      let h1 := raw_bytes[i + 1]!
      if let some hd1 := hexDigitToUInt8? h1 then
        if i + 2 < len then
          let h2 := raw_bytes[i + 2]!
          if let some hd2 := hexDigitToUInt8? h2 then
            -- decode the hex digits into a byte.
            (decoded.push (hd1 * 16 + hd2), i + 3)
          else
            -- not a valid second hex digit so keep the original bytes
            -- BUGBUG not working?
            (decoded.push c ++ decoded.push h1 ++ decoded.push h2, i + 3)
        else
          -- hit end of string, there is no h2.
          -- BUGBUG not working?
          (decoded.push c ++ decoded.push h1, i + 2)
      else
        -- not a valid hex digit so keep the original bytes
        (decoded.push c ++ decoded.push h1, i + 2)
    else
      (decoded.push c, i + 1)
    decoded := result
    i := pos
  return String.fromUTF8Unchecked decoded
where hexDigitToUInt8? (c : UInt8) : Option UInt8 :=
  if Zero ≤ c ∧ c ≤ Nine then some (c - Zero)
  else if Lettera ≤ c ∧ c ≤ Letterf then some (c - Lettera + 10)
  else if LetterA ≤ c ∧ c ≤ LetterF then some (c - LetterA + 10)
  else none

#eval decodeUri "%3a"            -- ":"    correct
#eval decodeUri "%3a%"         -- ":%" correct
#eval decodeUri "%3a%3"       -- ":%:3" weirdly broken, looks like went backwards
#eval decodeUri "%3a%3x"     -- ":%:3:x" weirdly broken
#eval decodeUri "%3a%3a%3a" -- ":::" correct

Is there a bug in my code here?
```

Reply:
Ok, I switched to this `(((decoded.push c).push h1).push h2, i + 3)` and got a bit of a speed up, so I pushed these fixes.

Reply:
@Kha - I realize it says unchecked, but how unsafe is it? the following seems to hang my VS code LSP server:

```lean
#eval String.fromUTF8Unchecked (ByteArray.mk #[0xab, 0xea])
```
did it walk into never never land looking for the rest of an incomplete UTF-8 sequence?

Reply:
@Kha I found a python test library here https://github.com/python/cpython/blob/main/Lib/test/test_urllib.py and copied many of the tests and fixed a bunch of bugs.

Reply:
I also fixed the snake casing problems.

Reply:
> @Kha - I realize it says unchecked, but how unsafe is it? the following seems to hang my VS code LSP server:
> 
> ```lean
> #eval String.fromUTF8Unchecked (ByteArray.mk #[0xab, 0xea])
> ```
> 
> did it walk into never never land looking for the rest of an incomplete UTF-8 sequence?

Oh that's funny, it looks like it overshoots the following condition, trapping itself in an infinite loop: https://github.com/leanprover/lean4/blob/185829d089080b290c012adb380beddfd49a4712/src/Init/Data/String/Basic.lean#L312

Reply:
should a file a separate bug on this, or do you want me to fix it in this PR?  If the latter, what is the fix exactly?

Reply:
I filed a [new bug](https://github.com/leanprover/lean4/issues/1470) so we don't lose it.